pkg/expr/nodes.go

```diff
@@ -242,15 +243,19 @@ func (dn *DSNode) Execute(ctx context.Context, now time.Time, _ mathexp.Vars, s
 
 	resp, err := s.dataService.QueryData(ctx, req)
 	if err != nil {
-		return mathexp.Results{}, err
+		return mathexp.Results{}, QueryError{
+			RefID:         dn.refID,
+			DatasourceUID: dn.datasource.UID,
+			Err:           err,
+		}
 	}
 
 	var result mathexp.Results
-	responseType, result, err = queryDataResponseToResults(ctx, resp, dn.refID, dn.datasource.Type, s)
+	responseType, result, err = queryDataResponseToResults(ctx, resp, dn.refID, dn.datasource.Type, dn.datasource.UID, s)
 	return result, err
 }
 
-func queryDataResponseToResults(ctx context.Context, resp *backend.QueryDataResponse, refID string, datasourceType string, s *Service) (string, mathexp.Results, error) {
+func queryDataResponseToResults(ctx context.Context, resp *backend.QueryDataResponse, refID string, datasourceType string, datasourceUid string, s *Service) (string, mathexp.Results, error) {
 	vals := make([]mathexp.Value, 0)
 	response, ok := resp.Responses[refID]
 	if !ok {
```

Comment:
It seems odd to me that we return no-data when the refID is not found -- this seems like it would be an error to me?

It is a bit separate, but there are a fair amount of things being passed to this func now, so my think was that the get response from responses  and checking for response.Error could be moved up to Execute.

Reply:
In ML PR I use this function and I will have to duplicate it there as well.

Reply:
I believe so, the Response should exist, even if it doesn't have any Frames

https://github.com/grafana/grafana-plugin-sdk-go/blob/7dd9759f51c13fec74c40e69ef20875c69f13050/backend/data.go#L15-L25